realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/SyncedRealmMigrationTests.kt

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import io.realm.SyncTestUtils.Companion.createTestUser
+import io.realm.entities.IndexedFields
+import io.realm.entities.PrimaryKeyAsString
+import io.realm.entities.StringOnly
+import io.realm.internal.OsObjectSchemaInfo
+import io.realm.internal.OsRealmConfig
+import io.realm.internal.OsSchemaInfo
+import io.realm.internal.OsSharedRealm
+import org.hamcrest.CoreMatchers
+import org.junit.*
+import org.junit.rules.ExpectedException
+import org.junit.Assert.*
+import org.junit.runner.RunWith
+import java.util.*
+import kotlin.test.assertFailsWith
+
+/**
+ * Testing methods around migrations for Realms using a [SyncConfiguration].
+ */
+@RunWith(AndroidJUnit4::class)
+class SyncedRealmMigrationTests {
+
+    @get:Rule
+    val configFactory = TestSyncConfigurationFactory()
+
+    @get:Rule
+    val thrown = ExpectedException.none()
```

Comment:
I think this Rule can largely be replaced with `assertFailsWith` ?

Reply:
Yes, I have it stashed somewhere as my unit tests of the implementation failed as it was not possible to `assertFailsWith` on an `AssertError`. I will give it another try if you also see the potential.

Reply:
Added a `assertFailsWithMessage<T>(Matcher<in String?>, () -> Unit)` in KotlinTestUtils